### PR TITLE
MC-39508: Spreading multiple GraphQL fragments into CategoryTree causes Magento to throw

### LINKED ISSUE #31086: Spreading multiple GraphQL fragments into CategoryTree causes Magento to throw
### Preconditions (*)
1. Magento 2.4.1 and 2.4-develop

### Steps to reproduce (*)
1. Run the following query:
```graphql
query GetCategoryTree($filters: CategoryFilterInput!) {
  categoryList(filters: $filters) {
    ...recursiveCategoryNode
  }
}

fragment recursiveCategoryNode on CategoryTree {
  ...categoryNode
  children {
    ...categoryNode
  }
}

fragment categoryNode on CategoryTree {
  id
}
```
{code}

I used the following argument:
{code:json}
```json
{
    "filters": {
        "ids": {
            "eq": "2"
        }
    }
}
```
{code}

I did not test on sample data but you can probably infer an analogous argument that matches the sample data. It seems that this only fails when there are categories that match the filter.

### Expected result

I expect to get a GraphQL response with category IDs.

### Actual result 
I get the following response:
{code:json}
{
    "errors": [
        {
            "debugMessage": "Argument 2 passed to Magento\\CatalogGraphQl\\Model\\Category\\DepthCalculator::calculate() must be an instance of GraphQL\\Language\\AST\\FieldNode, instance of GraphQL\\Language\\AST\\FragmentSpreadNode given, called in /var/www/html/vendor/magento/module-catalog-graph-ql/Model/Category/DepthCalculator.php on line 41",
            "message": "Internal server error",
            "extensions": {
                "category": "internal"
            },
            "locations": [
                {
                    "line": 2,
                    "column": 3
                }
            ],
            "path": [
                "categoryList"
            ]
        }
    >,
    "data": {
        "categoryList": null
    }
}
{code}

## 
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- < ] Severity: **S0** *- Affects critical data or functionality and leaves users without workaround.*
- [ ] Severity: **S1** *- Affects critical data or functionality and forces users to employ a workaround.*
- [x] Severity: **S2** *- Affects non-critical data or functionality and forces users to employ a workaround.*
- [ ] Severity: **S3** *- Affects non-critical data or functionality and does not force users to employ a workaround.*
- [ > Severity: **S4** *- Affects aesthetics, professional look and feel, “quality” or “usability”.*

Note that the following queries *do not* throw this error and complete as expected:

#### Using Two Fragments but Not Spreading into Root of `recursiveCategoryNode`
{code:graphql}
query GetCategoryTree($filters: CategoryFilterInput!) {
  categoryList(filters: $filters) {
    ...recursiveCategoryNode
  }
}

fragment recursiveCategoryNode on CategoryTree {
  id
  children {
    ...categoryNode
  }
}

fragment categoryNode on CategoryTree {
  id
}
{code}

and

#### Spreading into Root but Only Using One Fragment
{code:graphql}
query GetCategoryTree($filters: CategoryFilterInput!) {
  categoryList(filters: $filters) {
    ...categoryNode
    children {
        ...categoryNode
    }
  }
}

fragment categoryNode on CategoryTree {
    id
}
```

### PR DESCRIPTION
**Pull request by zakdma**
### Fixed Issues (if relevant)

1. Fixes https://github.com/magento/magento2/issues/31086

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by zakdma** [comment]
@magento run all tests

**Comment by zakdma** [comment]
@magento run Unit Tests

**Comment by zakdma** [comment]
@magento run Functional Tests EE

File: app/code/Magento/CatalogGraphQl/Model/Category/DepthCalculator.php [hunk 1]
```diff
@@ -26,22 +27,35 @@ class DepthCalculator
...
+    }
+
+    /**
+     * Calculate recursive the total depth of a category tree inside a GraphQL request
+     *
+     * @param ResolveInfo $resolveInfo
+     * @param Node $node
+     * @return int
+     */
+    private function calculateRecursive(ResolveInfo $resolveInfo, Node $node) : int
```
Comment (eduard13) [inline app/code/Magento/CatalogGraphQl/Model/Category/DepthCalculator.php]:
```suggestion
    private function calculateRecursive(ResolveInfo $resolveInfo, Node $node): int
```

File: dev/tests/api-functional/testsuite/Magento/GraphQl/Catalog/CategoryListTest.php [hunk 2]
```diff
@@ -842,4 +842,49 @@ public function testFilterCategoryNamedFragment()
         $this->assertEquals($result['categoryList'][0]['uid'], base64_encode('6'));
         $this->assertEquals($result['categoryList'][0]['url_path'], 'category-2');
     }
+
+    /**
+     * Test when there is recursive category node fragment
+     *
+     * @magentoApiDataFixture Magento/Catalog/_files/categories.php
+     */
+    public function testFilterCategoryRecursiveFragment() : void
```
Comment (eduard13) [inline dev/tests/api-functional/testsuite/Magento/GraphQl/Catalog/CategoryListTest.php]:
```suggestion
    public function testFilterCategoryRecursiveFragment(): void
```

File: app/code/Magento/CatalogGraphQl/Model/Category/DepthCalculator.php [hunk 3]
```diff
@@ -7,6 +7,7 @@
 
 namespace Magento\CatalogGraphQl\Model\Category;
 
+use GraphQL\Language\AST\Node;
```
Comment (eduard13) [inline app/code/Magento/CatalogGraphQl/Model/Category/DepthCalculator.php]: Please sort the imports

File: app/code/Magento/CatalogGraphQl/Model/Category/DepthCalculator.php [hunk 4]
```diff
@@ -26,22 +27,35 @@ class DepthCalculator
      */
     public function calculate(ResolveInfo $resolveInfo, FieldNode $fieldNode) : int
```
Comment (eduard13) [inline app/code/Magento/CatalogGraphQl/Model/Category/DepthCalculator.php]:
```suggestion
    public function calculate(ResolveInfo $resolveInfo, FieldNode $fieldNode): int
```

**Review by eduard13 — CHANGES_REQUESTED** [review]
Hi @zakdma, could you please check the minor suggestions below?

Thank you.

**Comment by eduard13** [comment]
@magento run Functional Tests EE, Integration Tests

**Comment by zakdma** [comment]
@magento run all tests

**Comment by eduard13** [comment]
@magento run Functional Tests B2B, WebAPI Tests

**Comment by eduard13** [comment]
@magento run Functional Tests B2B, Integration Tests

**Review by eduard13 — APPROVED** [review]
The failed Functional Tests B2B don't seem to be related to the PR's changes.

**Comment by magento-engcom-team** [comment]
Hi @eduard13, thank you for the review.
[ENGCOM-8935](https://jira.corp.magento.com/browse/ENGCOM-8935) has been created to process this Pull Request

**Review by engcom-Kilo — APPROVED** [review]

**Comment by dmitriyprime** [comment]
:heavy_check_mark: QA passed
**Before:**
:x: Internal server error:
`Argument 2 passed to Magento\CatalogGraphQl\Model\Category\DepthCalculator::calculate() must be an instance of GraphQL\Language\AST\FieldNode, instance of GraphQL\Language\AST\FragmentSpreadNode given, called in /var/www/html/magento2dev/app/code/Magento/CatalogGraphQl/Model/Category/DepthCalculator.php on line 41`
Screenshot from 2021-03-18 14-46-55


**After:**
:heavy_check_mark: Correct response with category IDs
Screenshot from 2021-03-18 14-43-53

**Comment by eduard13** [comment]
Hi @zakdma, looks like your changes were already merged internally, without the mentioned small changes. So, I'm closing this PR since the changes aren't needed anymore.
Please re-open it and correct me if I'm wrong.

Thank you.

**Pull request closed without merging**